src/util/expr.cpp

```diff
@@ -325,3 +326,182 @@ const_unique_depth_iteratort exprt::unique_depth_cbegin() const
 { return const_unique_depth_iteratort(*this); }
 const_unique_depth_iteratort exprt::unique_depth_cend() const
 { return const_unique_depth_iteratort(); }
+
+/// We use the precendences that most readers expect
+/// (i.e., the ones you learn in primary school),
+/// and stay clear of the surprising ones that C has.
+static bool bracket_subexpression(const exprt &sub_expr, const exprt &expr)
+{
+  // no need for parentheses whenever the subexpression
+  // doesn't have operands
+  if(!sub_expr.has_operands())
+    return false;
+
+  // * and / bind stronger than + and -
+  if(
+    (sub_expr.id() == ID_mult || sub_expr.id() == ID_div) &&
+    (expr.id() == ID_plus || expr.id() == ID_minus))
+    return false;
+
+  // ==, !=, <, <=, >, >= bind stronger than && and ||
+  if(
+    (sub_expr.id() == ID_equal || sub_expr.id() == ID_notequal ||
+     sub_expr.id() == ID_lt || sub_expr.id() == ID_gt ||
+     sub_expr.id() == ID_le || sub_expr.id() == ID_ge) &&
+    (expr.id() == ID_and || expr.id() == ID_or))
+    return false;
+
+  return true;
+}
+
+/// This formats turns a multi-ary expression,
+/// adding parentheses where indicated by the precedence.
+static std::ostream &format_rec(std::ostream &os, const multi_ary_exprt &src)
+{
+  bool first = true;
+
+  for(const auto &op : src.operands())
+  {
+    if(first)
+      first = false;
+    else
+      os << ' ' << src.id() << ' ';
+
+    const bool need_parentheses = bracket_subexpression(op, src);
+
+    if(need_parentheses)
+      os << '(';
+
+    os << format(op);
+
+    if(need_parentheses)
+      os << ')';
+  }
+
+  return os;
+}
+
+/// This formats a binary expression
+static std::ostream &format_rec(std::ostream &os, const binary_exprt &src)
+{
+  return format_rec(os, to_multi_ary_expr(src));
+}
+
+/// This formats a unary expression,
+/// adding parentheses very aggressively.
+static std::ostream &format_rec(std::ostream &os, const unary_exprt &src)
+{
+  if(src.id() == ID_not)
+    os << '!';
+  else if(src.id() == ID_unary_minus)
+    os << '-';
+  else
+    return os << src.pretty();
+
+  if(src.op0().has_operands())
+    return os << '(' << format(src.op0()) << ')';
+  else
+    return os << format(src.op0());
+}
+
+static std::ostream &format_rec(std::ostream &os, const constant_exprt &src)
```

Comment:
The others are documented, so please document it briefly as well.

Reply:
done